### ISSUE TITLE
error : undefined reference to `PROV_DEVICE_REG_STATUSStrings' and error : undefined reference to `PROV_DEVICE_RESULTStrings'

### ISSUE THREAD

**Issue by apostrophesoftware** [Azure/azure-iot-sdk-c#1505]
**Development Machine, OS, Compiler (and Other Relevant Toolchain Info)**

PRETTY_NAME="Raspbian GNU/Linux 10 (buster)"
NAME="Raspbian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"

gcc (Raspbian 8.3.0-6+rpi1) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

**SDK Version (Please Give Commit SHA if Manually Compiling)**

Release 2020-01-22
built using -Duse_prov_client:BOOL=ON

**Protocol**

MQTT

**Describe the Bug**
When linking I get the the following error
`error : undefined reference to 'PROV_DEVICE_REG_STATUSStrings'`

I am linking using the following libraries:

`iothub_client
prov_device_client
iothub_client_mqtt_transport
umqtt
prov_mqtt_transport
prov_mqtt_ws_transport
aziotsharedutil
prov_auth_client
hsm_security_client
prov_device_ll_client
msr_riot
utpm
parson`

`nm libprov_device_client.a` shows the following

`prov_device_client.c.o:
         U free
00000418 r __func__.7487
0000042c r __func__.7496
00000448 r __func__.7506
0000045c r __func__.7519
00000470 r __func__.7539
0000048c r __func__.7556
000004a4 r __func__.7575
000004cc r __func__.7583
         U get_time
         U Lock
         U Lock_Deinit
         U Lock_Init
         U malloc
         U memset
00000040 d OPTION_ADDRESS_TYPE
00000044 d OPTION_ADDRESS_TYPE_DOMAIN_SOCKET
00000048 d OPTION_ADDRESS_TYPE_IP_SOCKET
0000002c d OPTION_CURL_FORBID_REUSE
00000028 d OPTION_CURL_FRESH_CONNECT
00000020 d OPTION_CURL_LOW_SPEED_LIMIT
00000024 d OPTION_CURL_LOW_SPEED_TIME
00000030 d OPTION_CURL_VERBOSE
00000070 d OPTION_DO_WORK_FREQUENCY_IN_MS
00000000 d OPTION_HTTP_PROXY
00000004 d OPTION_HTTP_TIMEOUT
00000034 d OPTION_NET_INT_MAC_ADDRESS
0000000c d OPTION_OPENSSL_CIPHER_SUITE
00000038 d OPTION_SET_TLS_RENEGOTIATION
0000003c d OPTION_TLS_VERSION
00000008 d OPTION_TRUSTED_CERT
00000018 d OPTION_X509_ECC_CERT
0000001c d OPTION_X509_ECC_KEY
00000060 d PROV_API_VERSION
0000004c d PROV_ASSIGNED_STATUS
00000050 d PROV_ASSIGNING_STATUS
000001fc T Prov_Device_Create
0000040c T Prov_Device_Destroy
00000ab8 T Prov_Device_Get_Provisioning_Payload
000009ec T Prov_Device_GetVersionString
         U Prov_Device_LL_Create
         U Prov_Device_LL_Destroy
         U Prov_Device_LL_DoWork
         U Prov_Device_LL_Get_Provisioning_Payload
         U Prov_Device_LL_GetVersionString
         U Prov_Device_LL_Register_Device
         U Prov_Device_LL_SetOption
         U Prov_Device_LL_Set_Provisioning_Payload
000005e0 T Prov_Device_Register_Device
000007cc T Prov_Device_SetOption
00000a04 T Prov_Device_Set_Provisioning_Payload
00000058 d PROV_DISABLE_STATUS
00000068 d PROV_OPTION_LOG_TRACE
0000006c d PROV_OPTION_TIMEOUT
00000064 d PROV_REGISTRATION_ID
00000054 d PROV_UNASSIGNED_STATUS
0000005c d RETRY_AFTER_KEY_VALUE
00000000 t ScheduleWork_Thread
         U srand
00000120 t StartWorkerThreadIfNeeded
         U strcmp
00000010 d SU_OPTION_X509_CERT
00000014 d SU_OPTION_X509_PRIVATE_KEY
         U ThreadAPI_Create
         U ThreadAPI_Exit
         U ThreadAPI_Join
         U ThreadAPI_Sleep
         U Unlock
         U xlogging_get_log_function`

`nm libprov_device_ll_client.a` shows the following:

`prov_device_ll_client.c.o:
         U Azure_Base64_Decode
         U BUFFER_delete
         U BUFFER_length
         U BUFFER_u_char
00001f78 t cleanup_prov_info
000008b8 t construct_security_type_json
00002628 t destroy_instance
         U difftime
         U free
00000ebc r __func__.7654
00000ee0 r __func__.7701
00000ef8 r __func__.7717
00000f10 r __func__.7732
00000f30 r __func__.7765
00000f54 r __func__.7800
00000f78 r __func__.7897
00000f98 r __func__.7921
00000fac r __func__.7948
00000fc4 r __func__.7979
00000fe4 r __func__.8016
00000ffc r __func__.8051
00001018 r __func__.8094
00001040 r __func__.8106
         U get_time
000000a8 d JSON_CUSTOM_DATA_TAG
000000b8 d JSON_ENDORSMENT_KEY_NODE
         U json_free_serialized_string
0000008c d JSON_NODE_ASSIGNED_HUB
0000007c d JSON_NODE_AUTH_KEY
00000094 d JSON_NODE_DATE_TIME
00000080 d JSON_NODE_DEVICE_ID
0000009c d JSON_NODE_ERROR_CODE
00000098 d JSON_NODE_ERROR_MSG
00000084 d JSON_NODE_KEY_NAME
00000088 d JSON_NODE_OPERATION_ID
00000078 d JSON_NODE_REG_STATUS
000000ac d JSON_NODE_RETURNED_DATA
00000074 d JSON_NODE_STATUS
00000090 d JSON_NODE_TPM_NODE
         U json_object_get_object
         U json_object_get_value
         U json_object_set_string
         U json_object_set_value
         U json_parse_string
         U json_serialize_to_string
000000bc d JSON_STORAGE_ROOT_KEY_NODE
         U json_value_free
         U json_value_get_number
         U json_value_get_object
         U json_value_get_string
         U json_value_init_object
         U malloc
         U mallocAndStrcpy_s
         U memset
00000468 t on_transport_error
00002048 t on_transport_registration_data
00002380 t on_transport_status
00000040 d OPTION_ADDRESS_TYPE
00000044 d OPTION_ADDRESS_TYPE_DOMAIN_SOCKET
00000048 d OPTION_ADDRESS_TYPE_IP_SOCKET
0000002c d OPTION_CURL_FORBID_REUSE
00000028 d OPTION_CURL_FRESH_CONNECT
00000020 d OPTION_CURL_LOW_SPEED_LIMIT
00000024 d OPTION_CURL_LOW_SPEED_TIME
00000030 d OPTION_CURL_VERBOSE
00000000 d OPTION_HTTP_PROXY
00000004 d OPTION_HTTP_TIMEOUT
00000070 d OPTION_LOG_TRACE
00000034 d OPTION_NET_INT_MAC_ADDRESS
0000000c d OPTION_OPENSSL_CIPHER_SUITE
00000038 d OPTION_SET_TLS_RENEGOTIATION
0000003c d OPTION_TLS_VERSION
00000008 d OPTION_TRUSTED_CERT
00000018 d OPTION_X509_ECC_CERT
0000001c d OPTION_X509_ECC_KEY
00000060 d PROV_API_VERSION
0000004c d PROV_ASSIGNED_STATUS
00000050 d PROV_ASSIGNING_STATUS
         U prov_auth_construct_sas_token
         U prov_auth_create
         U prov_auth_destroy
         U prov_auth_get_alias_key
         U prov_auth_get_certificate
         U prov_auth_get_endorsement_key
         U prov_auth_get_registration_id
         U prov_auth_get_storage_key
         U prov_auth_get_type
         U prov_auth_import_key
         U prov_auth_set_registration_id
000000a4 d PROV_BLACKLISTED_STATUS
000026f4 T Prov_Device_LL_Create
00002bb0 T Prov_Device_LL_Destroy
000033c8 T Prov_Device_LL_DoWork
000041a8 T Prov_Device_LL_Get_Provisioning_Payload
00004034 T Prov_Device_LL_GetVersionString
00002be0 T Prov_Device_LL_Register_Device
00003980 T Prov_Device_LL_SetOption
00004058 T Prov_Device_LL_Set_Provisioning_Payload
00000058 d PROV_DISABLE_STATUS
000000a0 d PROV_FAILED_STATUS
00000068 d PROV_OPTION_LOG_TRACE
0000006c d PROV_OPTION_TIMEOUT
00000064 d PROV_REGISTRATION_ID
00000000 t prov_transport_challenge_callback
00000d60 t prov_transport_create_json_payload
0000113c t prov_transport_process_json_reply
00000054 d PROV_UNASSIGNED_STATUS
000000b4 d REGISTRATION_ID
00000644 t retrieve_json_number
000006fc t retrieve_json_payload
00000758 t retrieve_json_string
000004fc t retrieve_status_type
0000005c d RETRY_AFTER_KEY_VALUE
000000b0 d SAS_TOKEN_SCOPE_FMT
         U sprintf
         U srand
         U strcmp
         U STRING_c_str
         U STRING_delete
         U strlen
00000010 d SU_OPTION_X509_CERT
00000014 d SU_OPTION_X509_PRIVATE_KEY
         U tickcounter_create
         U tickcounter_destroy
         U tickcounter_get_current_ms
         U URL_EncodeString
         U xlogging_get_log_function`

If I build for the SDK for Windows using the same switch `PROV_DEVICE_REG_STATUS` is defined in `prov_device_client.lib` but I am unable to find the library where `PROV_DEVICE_RESULT` is defined

On Windows, the prov_dev_client_sample builds successfully

What am I missing in the Linux environment?

**Comment by jbobotek** [comment]
This is interesting. On further inspection of our tests, it looks like we deactivate prov_client. It's possible we don't formally support prov_client for raspberry pi. Will investigate.

**Comment by apostrophesoftware** [comment]
Thanks, I will await the outcome of your investigation

**Comment by jbobotek** [comment]
So I built it with prov_client on, and the SDK still builds and passes tests. I also looked at the sample code and CmakeLists.txt, and I don't see anything that should exclude the files needed. PROV_DEVICE_RESULT is defined in the sample itself. I think it's an issue with your linking. If using CMake, I advise following a similar build strategy to this file: https://github.com/Azure/azure-iot-sdk-c/blob/master/jenkins/raspberrypi_c_buster.sh (Note, it is cross compiling, but a lot of the parameters can just be linked to /bin resources etc)

**Comment by ewertons** [comment]
Hi @apostrophesoftware , 
we will close this issue, but if you would like it to be reopened and follow up on this matter, please post a comment requesting so.

Thanks,
Azure IoT SDK Team.

**Comment by az-iot-builder-01** [comment]
@apostrophesoftware, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by JustElectron** [comment]
I just encountered the same issue.
My assumption is that it is caused by the compiler. I'm using the same SDK version for an new and old product.
For the new product (new compiler) no issue, but for the old product (old compiler) i got the same issue.
When compiling the lib, we had to disable werror, so that might be the issue.

I didn't find a fix to this, but i found out that the error is caused by `MU_ENUM_TO_STRING(PROV_DEVICE_REG_STATUS` and `MU_ENUM_TO_STRING(PROV_DEVICE_RESULT,`. Removing these removed the error for me.
